src/com/ms/silverking/cloud/dht/daemon/storage/convergence/ActiveRegionSync.java

```diff
@@ -34,411 +35,416 @@
-                
-                if (debug) {
-                    Log.warningf("%s", entry.toString());
-                }
-                if (srr != null) {
-                    srr.outstandingKeys.remove(entry);
-                }
-                try {
-                    svp = StorageValueAndParameters.createSVP(entry);
-                } catch (RuntimeException re) {
-                    Log.warningf("Exception processing %s", IPAddrUtil.addrToString(message.getOriginator()));
-                    throw re;
-                }
-                if (svp != null) {
-                    svpList.add(svp);
-                }
-            }
-            if (!svpList.isEmpty()) {
-                if (nsStore != null) {
-                    // FUTURE - support migration of user data
-                    nsStore.put(svpList, emptyUserData, this);
-                }
+  }
+
+  ////////////////////////////////////////////////////////////////////
+
+  private void setComplete() {
+    completionLock.lock();
+    try {
+      Log.warningAsyncf("ars setComplete %s", uuid);
+      isComplete = true;
+      completionCV.signalAll();
+    } finally {
+      completionLock.unlock();
+    }
+  }
+
+  public boolean waitForCompletion(long time, TimeUnit unit) {
+    do {
+      Timer timer;
+
+      timer = new SimpleTimer(TimeUnit.MILLISECONDS, timeoutCheckMillis);
+      completionLock.lock();
+      try {
+        while (!isComplete) {
+          boolean checkTimedOut; // just this check; not the sync
+
+          try {
+            if (debug || verbose) {
+              Log.warningAsyncf("ars completionCV.await %s", uuid);
             }
-            if (srr != null && srr.outstandingKeys.isEmpty()) {
-                Log.warningAsyncf("ars %s complete-srr %s", uuid, srr.getUUID());
-                outstandingSyncRetrievalRequests.remove(srr.getUUID());
-                inprocessSyncRetrievalRequests.remove(srr.getUUID());
+            checkTimedOut = !timer.await(completionCV);
+            if (debug || verbose) {
+              Log.warningAsyncf("ars completionCV.await done %s", uuid);
             }
-            checkForCompletion();
-        } catch (RuntimeException re) {
-            Log.warningf("Exception in ars for ns %x", namespace);
-            throw re;
-        }
-    }    
-    
-    ////////////////////////////////////////////////////////////////////
-    
-    private void sendChecksumTreeRequest(ChecksumTreeRequest ctr) {
-        MessageGroup    mg;
-        
-        mg = new ProtoChecksumTreeRequestMessageGroup(uuid, namespace, ctr.getTargetCP(), ctr.getCurCP(),  
-                                                      mgBase.getMyID(), ctr.getRegion(), false).toMessageGroup(); 
-        if (verbose || debug) {
-            Log.warningAsyncf("%x requestChecksumTree: %s\t%s\t%s\t%s", namespace, ctr.getReplica(), ctr.getRegion(), ctr.getTargetCP(), ctr.getCurCP());
+          } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)